app/bundles/LeadBundle/Model/TagModel.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\LeadBundle\Model;
+
+use Mautic\CoreBundle\Model\FormModel;
+use Mautic\LeadBundle\Entity\Tag;
+use Mautic\LeadBundle\Event\TagEvent;
+use Mautic\LeadBundle\Form\Type\TagEntityType;
+use Mautic\LeadBundle\LeadEvents;
+use Symfony\Component\EventDispatcher\Event;
+use Symfony\Component\HttpKernel\Exception\MethodNotAllowedHttpException;
+
+/**
+ * Class TagModel
+ * {@inheritdoc}
+ */
+class TagModel extends FormModel
+{
+    /**
+     * {@inheritdoc}
+     *
+     * @return string
+     */
+    public function getRepository()
+    {
+        return $this->em->getRepository(Tag::class);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return string
+     */
+    public function getPermissionBase()
+    {
+        return 'lead:leads';
+    }
+
+    /**
+     * Get a specific entity or generate a new one if id is empty.
+     *
+     * @param $id
+     *
+     * @return null|object
+     */
+    public function getEntity($id = null)
+    {
+        if (is_null($id)) {
+            return new Tag();
+        }
+
+        return parent::getEntity($id);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param       $entity
+     * @param       $formFactory
+     * @param null  $action
+     * @param array $options
+     *
+     * @return mixed
+     *
+     * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException
+     */
+    public function createForm($entity, $formFactory, $action = null, $options = [])
+    {
+        if (!$entity instanceof Tag) {
```

Comment:
Do not like this in new code - this is the consequence of extending old model.
We should require typehints at function level.

Reply:
Yep, but it would require to rewrite Mautic :) or maybe :/